kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,305 @@
+                    let Term {
+                        exp: exp2,
+                        sign: sign2,
+                        degree: degree2,
+                    } = t;
+                    let degree = &degree * &degree2;
+                    let sign = sign == sign2;
+                    let exp = FoldingExp::Mul(exp.clone(), Box::new(exp2));
+                    Term { exp, sign, degree }
+                };
+                let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+                extract_terms(*e2.clone(), terms, &f);
+            }
+        }
+        e @ FoldingExp::UnnormalizedLagrangeBasis(_) => {
+            let term = f(Term::new(e, Zero));
+            terms.push(term)
+        }
+        FoldingExp::Cache(id, e) => {
+            let f = |t| {
+                let Term { exp, sign, degree } = t;
+                let exp = FoldingExp::Cache(id, Box::new(exp));
+                Term { exp, sign, degree }
+            };
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e, terms, &f);
+        }
+    }
+}
+
+///simplyfies the expression, and separates it into terms of different degree y
+pub fn folding_expression<C: Copy, F: Fn(&FeatureFlag) -> bool>(
+    exp: &Expr<C>,
+    flag_resolver: &F,
+) -> IntegratedFoldingExpr<C> {
+    let simplied = simplify_expression(exp, flag_resolver);
+    let mut terms = vec![];
+    extract_terms(simplied, &mut terms, &|t| t);
+    let mut integrated = IntegratedFoldingExpr::default();
+    for term in terms.into_iter() {
+        let Term { exp, sign, degree } = term;
+        let t = (exp, sign);
+        match degree {
+            Degree::Zero => integrated.degree_0.push(t),
+            Degree::One => integrated.degree_1.push(t),
+            Degree::Two => integrated.degree_2.push(t),
+        }
+    }
+    integrated
+}
+fn simplify_expression<C: Copy, F: Fn(&FeatureFlag) -> bool>(
+    exp: &Expr<C>,
+    flag_resolver: &F,
+) -> FoldingExp<C> {
+    match exp {
+        Expr::Constant(c) => FoldingExp::Constant(*c),
+        Expr::Cell(v) => {
+            let Variable { col, row } = v;
+            let col = match col {
+                Column::Witness(i) => FoldingColumn::Witness(*i),
+                Column::Index(i) => FoldingColumn::Index(*i),
+                Column::Coefficient(i) => FoldingColumn::Coefficient(*i),
+                _ => unreachable!(),
+            };
+            let cell = Box::new(FoldingExp::Cell(col));
+            let shift = Box::new(FoldingExp::Cell(FoldingColumn::Shift));
+            match row {
+                CurrOrNext::Curr => *cell,
+                CurrOrNext::Next => FoldingExp::Mul(cell, shift),
+            }
+        }
+        //may be better to replace by mul by 2
+        Expr::Double(e) => {
+            let e = simplify_expression(e, flag_resolver);
+            FoldingExp::Double(Box::new(e))
+        }
+        //may be better to replace by mul by itself
```

Comment:
I intuitively agree with both square and double -- I don't see how pushing them deeper into processing hierarchy is useful in terms of performance (and semantics wise it's better to transform them into `a+a` and `a*a` right here).